src/state.rs

```diff
@@ -48,23 +58,63 @@ impl State {
             scarb_toolchain: scarb_toolchain.clone(),
             db_swapper: AnalysisDatabaseSwapper::new(),
             tricks: Owned::new(tricks.into()),
-            diagnostics_controller: DiagnosticsController::new(notifier.clone()),
+            diagnostics_controller,
+            analysis_progress_controller,
             proc_macro_controller,
             project_controller: ProjectController::initialize(scarb_toolchain, notifier),
         }
     }
 
     pub fn snapshot(&self) -> StateSnapshot {
-        StateSnapshot {
+        Beacon::wrap(SnapshotInternal {
             db: self.db.snapshot(),
             open_files: self.open_files.snapshot(),
             config: self.config.snapshot(),
+        })
+    }
+}
+
+pub struct Beacon<T> {
```

Comment:
is there a specific reason why are you wrapping `T` here?

can't this be a valueless object that is kept as a field in the old `StateSnapshot`? It would still be dropped whenever `StateSnapshot` is

Reply:
1. I resigned from the "on-drop" behaviour (because of lack of information when the analysis actually finished on a thread vs. the thread was cancelled) 
2. I can do this without wrapping - no problem there i think (discussion [here](https://github.com/software-mansion/cairols/pull/44#discussion_r1890257077))

Reply:
Done

Reply:
>I resigned from the "on-drop" behaviour (because of lack of information when the analysis actually finished on a thread vs. the thread was cancelled)

Can we add a comment about it somewhere near that?

Reply:
Bcs I was a second person that asked to implement it on drop xD